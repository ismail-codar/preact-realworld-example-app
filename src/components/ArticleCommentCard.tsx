import { FunctionalComponent, h } from 'preact'
import { dateFilter } from '../utils/filters'
import { useRootState } from '../store'
import { DEFAULT_AVATAR } from '../store/constants'

interface ArticleCommentCardProps {
  comment: ArticleComment;
  onDelete?: (commentId: number) => void;
}

const ArticleCommentCard: FunctionalComponent<ArticleCommentCardProps> = (props) => {
  const { comment, onDelete } = props
  const [{ user }] = useRootState()

  const isMineComment = user?.username === comment.author.username

  if (!comment.author) return null

  return (
    <div className="card">
      <div className="card-block">
        <p className="card-text">{comment.body}</p>
      </div>
      <div className="card-footer">
        <a href={`/@${comment.author.username}`} className="comment-author">
          <img src={comment.author.image || DEFAULT_AVATAR} className="comment-author-img" />
        </a>
        &nbsp;
        <a href={`/@${comment.author.username}`} className="comment-author">{comment.author.username}</a>
        <span className="date-posted">{dateFilter(comment.createdAt)}</span>
        <span className="mod-options">
          {/* <i className="ion-edit" /> */}
          {isMineComment && (
            <i className="ion-trash-a" onClick={() => onDelete?.(comment.id)} />
          )}
        </span>
      </div>
    </div>
  )
}

export default ArticleCommentCard
