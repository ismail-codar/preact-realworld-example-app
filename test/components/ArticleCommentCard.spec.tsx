import { h } from 'preact';
import { fireEvent, render, screen, waitFor } from '@testing-library/preact';
import { rest } from 'msw';
import { setupServer } from 'msw/node';

import ArticleCommentCard from '../../src/components/ArticleCommentCard';
import useStore from '../../src/store';

jest.mock('../../src/store', () => jest.fn());

const comment: ArticleComment = {
	id: 1,
	body: 'foo',
	createdAt: '2009-02-19T00:00:00.000Z',
	updatedAt: '2008-05-19T00:00:00.000Z',
	author: {
		username: 'SmokeTest',
		following: false
	}
};

const articleSlug = '/foo';

test('should render correctly', () => {
	const { asFragment } = render(
		<ArticleCommentCard articleSlug={articleSlug} comment={comment} onDelete={jest.fn()} />
	);

	expect(asFragment()).toMatchSnapshot();
});

test("should not display trash icon in other's comment", async () => {
	((useStore as unknown) as jest.Mock).mockReturnValue({ username: 'SmokeTest' });
	render(<ArticleCommentCard articleSlug={articleSlug} comment={comment} onDelete={jest.fn()} />);

	expect(screen.getByLabelText('Delete comment')).toBeInTheDocument();
});

test('should delete comment when trash icon clicked', () => {
	const server = setupServer(
		rest.post(
			`https://conduit.productionready.io/api/articles/${articleSlug}/comments/${comment.id}`,
			(_req, res, ctx) => {
				return res(ctx.status(200));
			}
		)
	);
	server.listen();
	((useStore as unknown) as jest.Mock).mockReturnValue({ username: 'SmokeTest' });

	const onDelete = jest.fn();
	render(<ArticleCommentCard articleSlug={articleSlug} comment={comment} onDelete={onDelete} />);

	fireEvent.click(screen.getByLabelText('Delete comment'));
	waitFor(() => expect(onDelete).toHaveBeenCalled());

	server.close();
});
