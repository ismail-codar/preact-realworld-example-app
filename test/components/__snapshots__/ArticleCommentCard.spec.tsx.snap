// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="card-block"
    >
      <p
        class="card-text"
      >
        foo
      </p>
    </div>
    <div
      class="card-footer"
    >
      <a
        class="comment-author"
        href="/@SmokeTest"
      >
        <img
          alt="User's profile picture"
          class="comment-author-img"
          src="https://static.productionready.io/images/smiley-cyrus.jpg"
        />
      </a>
      Â 
      <a
        class="comment-author"
        href="/@SmokeTest"
      >
        SmokeTest
      </a>
      <span
        class="date-posted"
      >
        February 18, 2009
      </span>
      <span
        class="mod-options"
      />
    </div>
  </div>
</DocumentFragment>
`;
